vllm/v1/core/scheduler.py

```diff
@@ -330,6 +367,23 @@ def schedule(self) -> "SchedulerOutput":
                 self.kv_cache_manager.get_num_common_prefix_blocks(
                     any_request, len(self.running)))
 
+        # Prepare the guided decoding bitmask for this batch.
+        grammar_bitmask: Optional[np.ndarray] = None
+        if guided_decoding_request_ids:
+            # Fill the bitmask using the index of each request equal to its
+            # position in the batch. Resize the bitmask down to the size of
+            # the batch.
+            bitmask_tensor = self.guided_decoding_manager.grammar_bitmask
+            assert bitmask_tensor is not None
+            for req_id, batch_index in guided_decoding_request_ids.items():
+                request = self.requests[req_id]
+                assert request.grammar is not None
+                if not request.grammar.matcher.is_terminated():
+                    request.grammar.fill_bitmask(bitmask_tensor, batch_index)
+            if len(self.running) < bitmask_tensor.shape[0]:
+                bitmask_tensor = bitmask_tensor[:len(self.running)]
+            grammar_bitmask = bitmask_tensor.numpy()
```

Comment:
I think this shouldn't be in the `schedule` method nor in the scheduler code. This logic should be under the `guided_decoding` folder (something like `guided_decoding/utils.py`).

Reply:
The above 15 lines or so are now replaced with this:

```python
        grammar_bitmask = self.guided_decoding_manager.grammar_bitmask(
            self.requests, guided_decoding_request_ids, len(self.running))
```